### PR TITLE
Update to latest nightly, removing LocalWaker and introducing ArcWake

### PR DESCRIPTION
**Pull request by tinco**
I didn't really delve into the implications of these changes, I noticed the project no
longer compiled on nightly, and found this PR:

https://github.com/rust-lang/rust/pull/57992

So I basically mapped the new unit tests in there to this project, until
this project compiled again. Then I ran the suite and it passed so
I reckon it at least sort of works. It definitely needs a review
by someone who actually understands the code (hoping that is you ;))

### CODE REVIEW

File: src/arc_wake.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+// From htttps://github.com/rust-lang/rustc
```
Comment (tinco) [inline src/arc_wake.rs]: Since I pulled this file from rust-lang/rustc I included their LICENSE-MIT in the top of the file.

**Comment by de-vri-es** [comment]
Thanks for the PR! I'll need some time to properly review this, since wakers aren't the easiest thing in the world to do safely. I'd kinda like to avoid the inclusion of `ArcWaker`, but I'll have to get up to speed with the current state of futures first to see if that's feasible.

File: src/arc_wake.rs (hunk 1 above)
Comment (de-vri-es) [reply src/arc_wake.rs]: Very good :)  However, the license also needs to be shipped with binary releases. We're not hosting any of those here, so technically this repository is fine. But if people use this crate as dependency, they'll almost certainly miss the license if it's not in the LICENSE file.

Could you add the license to the LICENSE file instead, below the BSD license, with a note that it applies to this file?

**Review by de-vri-es — COMMENTED** [review]
I had some more time to review the PR. I believe it indeed restores functionality correctly.

It may actually be possible to map the `Waker` and `Notifier` traits to each-other without creating an `Arc`, but that was already on the TODO list anyway. So lets merge this with `Arc`, and I'll see it can be done without them later.

Just one reply to your comment regarding the license that we should sort out first.

File: src/arc_wake.rs (hunk 1 above)
Comment (de-vri-es) [reply src/arc_wake.rs]: The comment about the origin of the file is nice anyway, lets keep that in here.

**Comment by tinco** [comment]
Like this? :)

**Comment by tinco** [comment]
Oh jeez :P I'm at my work so I can't really do work on fixing this right now. Maybe in the weekend..

**Comment by de-vri-es** [comment]
Yeah the license is great like that. But it seems we've been caught up by rust development again :p

**Comment by de-vri-es** [comment]
I tried to fix it, which ironically seemed to require removing `ArcWake`. The vtable now has to be passed to RawWaker as `&'static`, which doesn't seem possible with a generic type argument (though I may have overlooked a way, I'm not sure).

Either way, I made a vtable for `TaskWaker` instead, which doesn't have any type arguments. Would you mind double-checking that I didn't mess anything up with leaking and dropping the boxed `TaskWaker`?

Also, my apolagies for pushing to your fork, I always get a bit paranoid when that happens. In case you didn't know, github allows that by default for collaborators of the target repository for a PR (but you can disable it).

File: src/futures_future.rs [hunk 2]
```diff
@@ -18,9 +18,34 @@ struct TaskWaker {
...
+}
+
+unsafe fn wake_task_waker(data: *const ()) {
+	let data = &*(data as *const TaskWaker);
+	data.task.notify();
+}
+
+impl TaskWaker {
+	unsafe fn into_raw_waker(task: futures::task::Task) -> RawWaker {
+		let data = Box::leak(Box::new(task)) as *const _ as *const ();
```
Comment (tinco) [inline src/futures_future.rs]: What's going on with the double `as ..`, is that necessary?

Comment (de-vri-es) [reply src/futures_future.rs]: Yeah, you can't cast directly to a pointer of a different type, but you can cast pointers to pointers of different types. So first it casts to a `*const futures::task::Task` (abbreviated as `*const _`, then to a `*const ()`.

**Comment by de-vri-es** [comment]
I'll go ahead and merge. Thanks for the PR, I hadn't even noticed it was broken yet :)

**Comment by de-vri-es** [comment]
heh, it already didn't compile anymore. Fixed in 940ccd948fddc0e6ac552c9d9196fd478f5e450c.

**Pull request merged**